Network cli interaction logging

bdowling opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Adds the interaction with the device to the debug output.

This is not quite what I desired in #29548 over a year ago, but it does help to debug some of the 'behind the scenes' cli chat with the device and when things go awry (#33794)

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (network_cli_interaction_logging 5cd847b998) last updated 2017/12/14 21:50:30 (GMT +000)
```

##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
2017-12-14 21:45:35,725 p=19870 u=bdowling |   19877 1513287935.71843: CLI-SEND: terminal length 0\r
2017-12-14 21:45:35,864 p=19870 u=bdowling |   19877 1513287935.86334: CLI-RECV: terminal length 0\r\nbd-csr1000v-00>
2017-12-14 21:45:35,865 p=19870 u=bdowling |   19877 1513287935.86452: CLI-SEND: terminal width 512\r
2017-12-14 21:45:36,022 p=19870 u=bdowling |   19877 1513287936.02179: CLI-RECV: terminal width 512\r\nbd-csr1000v-00>
2017-12-14 21:45:36,023 p=19870 u=bdowling |   19877 1513287936.02305: CLI-SEND: \r
2017-12-14 21:45:36,057 p=19870 u=bdowling |   19877 1513287936.05630: CLI-RECV: \r\nbd-csr1000v-00>
2017-12-14 21:45:36,058 p=19870 u=bdowling |   19877 1513287936.05765: CLI-SEND: enable\r
2017-12-14 21:45:36,103 p=19870 u=bdowling |   19877 1513287936.10237: CLI-PROMPT: Matched [\r\n]?password: $, sent '<answer>\r'
2017-12-14 21:45:36,168 p=19870 u=bdowling |   19877 1513287936.16800: CLI-RECV: enable\r\nPassword: \r\nbd-csr1000v-00#
2017-12-14 21:45:36,170 p=19870 u=bdowling |   19877 1513287936.16923: CLI-SEND: \r
2017-12-14 21:45:36,187 p=19870 u=bdowling |   19877 1513287936.18633: CLI-RECV: \r\nbd-csr1000v-00#
2017-12-14 21:45:36,633 p=19870 u=bdowling |   19877 1513287936.63297: CLI-SEND: show version\r
```

bdowling committed:
Make lower-level device interaction available as debug output

bdowling commented:
Oops, this does depend on #33049 but, but didn't intend to show that in the push.. hopefully that gets merged.

ganeshrn commented on lib/ansible/plugins/connection/network_cli.py:
```diff
@@ -294,10 +300,12 @@ def send(self, command, prompt=None, answer=None, newline=True, sendonly=False):
         '''
         try:
             self._history.append(command)
+            display.debug("CLI-SEND: %s\\r" % self._escape_crnl(command))
```
It is not recommended to log commands executed on the remote device due to security reasons.

ganeshrn commented on lib/ansible/plugins/connection/network_cli.py:
```diff
@@ -294,10 +300,12 @@ def send(self, command, prompt=None, answer=None, newline=True, sendonly=False):
             if sendonly:
                 return
             response = self.receive(command, prompt, answer, newline)
+            display.debug("CLI-RECV: %s" % self._escape_crnl(self._last_response))
```
Same as above

bdowling commented:
These are debug statements, only logged when `ANSIBLE_DEBUG=True`.  Someone using this needs this information to troubleshoot what went wrong and can be accountable for safety of said logging.  The actual CLI interaction with the network devices especially is very important for debugging what is going on behind the scenes.  There have been numerous tickets recently relating to extra CR/NL characters, non-synchronized prompts, confirmation prompts etc.  Without having access to the full device CLI interaction these things are very hard to troubleshoot.  These debug statements give much greater access to exactly how the interaction played out.

Additionally, network_cli runs on the ansible-server, where it is known what was run already, no "remote" is involved, if you can't trust the host that ansible is running on then you have larger problems.

bdowling committed:
Added showing the growing buffer that prompts are searched in

While this has a side effect of showing that reads are character by character, this also would
let someone see that a prompt could be matching mid-stream and the output of the logs more
informative.  Flushing the send buffer would alleviate this most of the time, but not yet
sure how that can be done with paramiko

ganeshrn commented:
@bdowling Completely agree with your observation that without access to device interaction it is difficult to debug if something goes wrong at the connection plugin level. However, logging the configuration command in plain text can result in leaking of sensitive information in log file without user knowing about it which is not recommended.

I have created a feature request to add support secure session tracing capability. Hence closing this PR.
https://github.com/ansible/ansible/issues/34997

Thank you for your contribution.

bdowling commented:
:( 

Device interaction logging should be a first-class citizen.  We already have capabilities to REDACT variables in logs that are marked as protected, that could be included here if you feel it is necessary.  I feel as though you are going at great lengths from preventing sophisticated users of a sophisticated tool from whacking their thumb with a hammer.  

As I mentioned over a year ago, providing this output to the user would go a long way to improving the acceptance of what ansible is actually doing on the network devices among network engineers.  Black Boxes are discouraged in network engineering.  Re: #5402

privateip commented:
@bdowling I have a design in mind to fix this in 2.6 and think we can get there.  This isn't about trying to prevent anyone for doing anything as much as it is an issue where by blindly logging all inputs and outputs opens a *huge* security hold that will cause Ansible to fail an one of a number of security audits.   

WRT the comment about the REDACT capabilities, there are two points here.  1) Where the logging would need to be inserted is well below where the heuristics happen in Ansible today so that cannot be leveraged.  2) The current heuristics in Ansible wouldn't catch some of the cases that would be exposed in a security audit.

I'm very much sensitive to the why logging is so desired (both from an operations standpoint and a developer / debugging standpoint) but we have to do this in a way that balances with the requirements of organizations that need Ansible to run in very secure environments.
